builder.go

```diff
@@ -0,0 +1,247 @@
+package biscuit
+
+import (
+	"errors"
+	"io"
+
+	"github.com/flynn/biscuit-go/datalog"
+	"github.com/flynn/biscuit-go/pb"
+	"github.com/flynn/biscuit-go/sig"
+	"google.golang.org/protobuf/proto"
+)
+
+var (
+	ErrDuplicateFact     = errors.New("biscuit: fact already exists")
+	ErrInvalidBlockIndex = errors.New("biscuit: invalid block index")
+)
+
+type Builder interface {
+	AddAuthorityFact(fact *Fact) error
+	AddAuthorityRule(rule *Rule) error
+	AddAuthorityCaveat(rule *Rule) error
+	AddRight(resource, right string) error
+	Build() (*Biscuit, error)
+}
+
+type builder struct {
+	rng  io.Reader
+	root sig.Keypair
+
+	symbolsStart int
+	symbols      *datalog.SymbolTable
+	facts        *datalog.FactSet
+	rules        []*datalog.Rule
+	caveats      []*datalog.Caveat
+	context      string
+}
+
+func NewBuilder(rng io.Reader, root sig.Keypair) Builder {
+	return BuilderWithSymbols(rng, root, DefaultSymbolTable)
+}
+
+func BuilderWithSymbols(rng io.Reader, root sig.Keypair, symbols *datalog.SymbolTable) Builder {
+	return &builder{
+		rng:  rng,
+		root: root,
+
+		symbolsStart: symbols.Len(),
+		symbols:      symbols,
+		facts:        new(datalog.FactSet),
+	}
+}
+
+func (b *builder) AddAuthorityFact(fact *Fact) error {
+	if len(fact.Predicate.IDs) == 0 {
+		fact.Predicate.IDs = []Atom{SymbolAuthority}
+	} else if fact.Predicate.IDs[0] != SymbolAuthority {
+		fact.Predicate.IDs[0] = SymbolAuthority
+	}
+
+	dlFact := fact.convert(b.symbols)
+	if !b.facts.Insert(dlFact) {
+		return ErrDuplicateFact
+	}
+
+	return nil
+}
+
+func (b *builder) AddAuthorityRule(rule *Rule) error {
+	if len(rule.Head.IDs) == 0 {
+		rule.Head.IDs = []Atom{SymbolAuthority}
+	} else if rule.Head.IDs[0] != SymbolAuthority {
+		rule.Head.IDs[0] = SymbolAuthority
+	}
+
+	dlRule := rule.convert(b.symbols)
+	b.rules = append(b.rules, &dlRule)
+	return nil
+}
+
+func (b *builder) AddAuthorityCaveat(rule *Rule) error {
+	b.caveats = append(b.caveats, &datalog.Caveat{Queries: []datalog.Rule{rule.convert(b.symbols)}})
+	return nil
+}
+
+func (b *builder) AddRight(resource, right string) error {
+	return b.AddAuthorityFact(&Fact{
+		Predicate: Predicate{
+			Name: "right",
+			IDs: []Atom{
+				SymbolAuthority,
+				String(resource),
+				Symbol(right),
+			},
+		},
+	})
+}
+
+func (b *builder) Build() (*Biscuit, error) {
+	return New(b.rng, b.root, b.symbols, &Block{
+		index:   0,
+		symbols: b.symbols.SplitOff(b.symbolsStart),
+		facts:   b.facts,
+		rules:   b.rules,
+		caveats: b.caveats,
+		context: b.context,
+	})
+}
+
+type Unmarshaler struct {
+	Symbols *datalog.SymbolTable
+}
+
+var DefaultUnmarshaler = &Unmarshaler{Symbols: DefaultSymbolTable}
```

Comment:
I'd unexport this and make a top-level `Unmarshal` function, that way the symbol table doesn't inadvertently get modified globally.